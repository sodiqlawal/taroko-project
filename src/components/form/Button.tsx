import classNames from "classnames";
import { FC, ButtonHTMLAttributes } from "react";
import styles from "./form.module.scss";

type ButtonProps = {
    content: string;
    loading?: boolean;
    isFull?: boolean;
    className?: string;
    isDanger?: boolean;
    isPrimary?: boolean;
    isMedium?: boolean;
    growth?: boolean;
    simple?: boolean
}

const Button: FC<
    ButtonHTMLAttributes<HTMLButtonElement> & ButtonProps
> = ({ content, className, loading, isFull, isDanger, isPrimary, isMedium, growth, simple, ...props }) => {
    return (
        <button
            type={props.type || "button"}
            className={classNames(styles.button, className, {
                [styles.full]: isFull,
                [styles.danger]: isDanger,
                [styles.primary]: isPrimary,
                [styles.medium]: isMedium,
                [styles.growth]: growth,
                [styles.simple]: simple,
                [styles.disabled]: props.disabled
            })}
            {...props}
        >
            {loading ? "loading..." : content}
        </button>
    );
};

export default Button;
