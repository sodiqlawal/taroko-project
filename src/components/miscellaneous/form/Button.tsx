import classNames from "classnames";
import { FC, ButtonHTMLAttributes } from "react";
import styles from "./form.module.scss";

const Button: FC<
    ButtonHTMLAttributes<HTMLButtonElement> & {
        content: string;
        loading?: boolean;
        isFull?: boolean;
        className?: string;
        isDanger?: boolean;
        isPrimary?: boolean;
        isMedium?: boolean;
        growth?: boolean;
        simple?:boolean
    }
> = ({ content, className, loading, isFull, ...props }) => {
    return (
        <button
            type={props.type || "button"}
            className={classNames(styles.button, className, {
                [styles.full]: isFull,
                [styles.danger]: props.isDanger,
                [styles.primary]: props.isPrimary,
                [styles.medium]: props.isMedium,
                [styles.growth]: props.growth,
                [styles.simple]: props.simple,
                [styles.disabled]: props.disabled
            })}
            {...props}
        >
            <span>{loading ? "loading..." : content}</span>
        </button>
    );
};

export default Button;
