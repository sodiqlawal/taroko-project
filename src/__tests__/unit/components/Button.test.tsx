import { render, screen, fireEvent } from "@testing-library/react";
import Button from "@/components/form/Button";

const defaultProps: {
    content: string;
    onClick: () => void;
    type: "submit" | "reset" | "button";
    disabled: boolean;
    loading: boolean
} = {
    content: "Submit",
    onClick: jest.fn(),
    type: "submit",
    disabled: false,
    loading: false
};

const setupButton = (props = defaultProps) => render(<Button {...props} />);

describe("The <SubmitButton /> component", () => {

    it("should match snapshot", () => {
        const { asFragment } = setupButton();

        expect(asFragment()).toMatchSnapshot()
    });

    it("should render the button and content", () => {
        setupButton();

        const content = screen.getByText(defaultProps.content);
        const loadingText = screen.queryByText(/loading.../);

        expect(content).toBeInTheDocument();
        expect(loadingText).not.toBeInTheDocument();
    });

    it("should call the click handler", () => {
        setupButton();

        const button = screen.getByRole("button", { name: /Submit/i });

        fireEvent.click(button);

        expect(defaultProps.onClick).toHaveBeenCalled();
    });

    it("should disable button if disabled is true", () => {
        defaultProps.disabled = true;
        setupButton();

        const button = screen.getByRole("button", { name: /Submit/i });

        expect(button).toBeDisabled();
    });

    it("should display `loading...` if loading is true", () => {
        defaultProps.loading = true;
        setupButton();

        const loadingText = screen.getByText(/loading.../);

        expect(loadingText).toBeInTheDocument();
    });
});
