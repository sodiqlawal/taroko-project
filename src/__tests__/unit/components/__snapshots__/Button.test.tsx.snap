// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <SubmitButton /> component should match snapshot 1`] = `
<DocumentFragment>
  <button
    class="button"
    type="submit"
  >
    Submit
  </button>
</DocumentFragment>
`;
